src/index/CompressedRelation.cpp

```diff
@@ -312,80 +341,140 @@ bool CompressedRelationReader::FilterDuplicatesAndGraphs::canBlockBeSkipped(
-
   if (relevantBlockMetadata.empty()) {
-    co_return;
+    return IdTableGeneratorInputRange{};
   }
 
-  // Some preparation.
-  auto [beginBlockMetadata, endBlockMetadata] =
-      getBeginAndEnd(relevantBlockMetadata);
-  LazyScanMetadata& details = co_await cppcoro::getDetails;
-  size_t numBlocksTotal = endBlockMetadata - beginBlockMetadata;
-  auto config =
-      getScanConfig(scanSpec, additionalColumns, locatedTriplesPerBlock);
+  struct Generator : ad_utility::InputRangeFromGet<IdTable, LazyScanMetadata> {
+    ScanSpecification scanSpec;
+    std::vector<CompressedBlockMetadata> relevantBlockMetadata;
+    ColumnIndices additionalColumns;
+    CancellationHandle cancellationHandle;
+    const LocatedTriplesPerBlock& locatedTriplesPerBlock;
+    LimitOffsetClause limitOffset;
+    ad_utility::InputRangeTypeErased<IdTable, LazyScanMetadata>
+        blockGenerator{};
+    bool firstBlockYielded{false};
+    bool middleBlocksYielded{false};
+    bool lastBlockYielded{false};
+    bool generatorCreated{false};
+    std::vector<CompressedBlockMetadata>::iterator beginBlockMetadata;
+    std::vector<CompressedBlockMetadata>::iterator endBlockMetadata;
+    const CompressedRelationReader* reader;
+    ScanImplConfig config;
+    IdTableGeneratorInputRange middleBlocksGenerator{};
+    // We will modify `limitOffset` as we go. We make a copy of the original
+    // value for some sanity checks at the end of the function.
+    const LimitOffsetClause originalLimit{limitOffset};
+    std::size_t numBlocksTotal;
+
+    Generator(ScanSpecification scanSpec_,
+              std::vector<CompressedBlockMetadata> relevantBlockMetadata_,
+              ColumnIndices additionalColumns_,
+              CancellationHandle cancellationHandle_,
+              const LocatedTriplesPerBlock& locatedTriplesPerBlock_,
+              LimitOffsetClause limitOffset_,
+              const CompressedRelationReader* reader_, ScanImplConfig config_)
+        : scanSpec{scanSpec_},
+          relevantBlockMetadata{std::move(relevantBlockMetadata_)},
+          additionalColumns{std::move(additionalColumns_)},
+          cancellationHandle{cancellationHandle_},
+          locatedTriplesPerBlock{locatedTriplesPerBlock_},
+          limitOffset{limitOffset_},
+          reader{reader_},
+          config{config_} {
+      beginBlockMetadata = ql::ranges::begin(relevantBlockMetadata);
+      endBlockMetadata = ql::ranges::end(relevantBlockMetadata);
+
+      numBlocksTotal = endBlockMetadata - beginBlockMetadata;
+    }
 
-  // Helper lambda for reading the first and last block, of which only a part
-  // is needed.
-  auto getIncompleteBlock = [&](auto it) {
-    auto result = readPossiblyIncompleteBlock(
-        scanSpec, config, *it, std::ref(details), locatedTriplesPerBlock);
-    cancellationHandle->throwIfCancelled();
-    return result;
-  };
+    auto getIncompleteBlock(std::vector<CompressedBlockMetadata>::iterator it) {
```

Comment:
add a `using` declaration to the beginning of this class for this commonly used `iterator` type.

Reply:
done in [this](https://github.com/ad-freiburg/qlever/pull/2236/425895079167b68669ef0ca0aba75275c452e426) commit